src/ngx_http_lua_socket_tcp.c

```diff
@@ -4915,13 +5003,28 @@ ngx_http_lua_socket_cleanup_compiled_pattern(lua_State *L)
 {
     ngx_http_lua_socket_compiled_pattern_t      *cp;
 
-    ngx_http_lua_dfa_edge_t         *edge, *p;
-    unsigned                         i;
+    ngx_http_lua_socket_tcp_upstream_t      *u;
+    ngx_http_lua_dfa_edge_t                 *edge, *p;
+    unsigned                                 i;
 
     dd("cleanup compiled pattern");
 
     cp = lua_touserdata(L, 1);
-    if (cp == NULL || cp->recovering == NULL) {
+    if (cp == NULL) {
+        return 0;
+    }
+
+    u = cp->upstream;
+    if (u != NULL) {
+        /* ensure request not released */
```

Comment:
what does this comment mean? could you please make it more clear?

`u->cleanup != NULL` means request is not released?

Reply:
If `socket` obj is freed before `pattern` obj, `ngx_http_lua_socket_tcp_finalize` will be called and set `cleanup` field to `NULL`.
I will try to make the comment more clear.

Reply:
It seems the inner `if` statement is meaningless, since `request` or `socket` has been freed before pattern and function `ngx_http_lua_socket_tcp_finalize` has set field `upstream` to [NULL](https://github.com/openresty/lua-nginx-module/pull/2145/files#diff-ebaa829579f91b18746b0bb13948131e7dcaa19dbae1d843c6fbbebc682951a2R4331)